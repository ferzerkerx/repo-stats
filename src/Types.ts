export interface MetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
  since?: Date;
  until?: Date;
}
