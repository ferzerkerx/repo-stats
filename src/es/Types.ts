import { MetricItem } from '../metrics';

export interface ElasticSearchService {
  push(payload: MetricItem, shouldReplaceEntry?: boolean): Promise<any>;
}
