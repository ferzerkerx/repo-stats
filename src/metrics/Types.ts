export interface CollectorService<
  T extends CollectorConfig,
  K extends MetricItem
> {
  fetch(config: T): Promise<K[]>;
}

export interface MetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
}

export interface CollectorConfig {}

export interface CollectorModuleFactory<
  T extends CollectorConfig,
  K extends MetricItem
> {
  collectorInstance(): CollectorService<T, K>;
  collectorConfiguration(obj: any): T;
}
