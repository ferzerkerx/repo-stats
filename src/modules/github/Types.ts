export interface GithubEntry {
  sha: string;
  createdAt: Date;
  author: string;
  filename: string;
  additions: number;
  deletions: number;
  categories?: string[];
}

export interface GithubCommitFile {
  filename: string;
  additions: number;
  deletions: number;
}

export class GithubCommit {
  sha: string;
  createdAt: Date;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  message: string;
  files: GithubCommitFile[];
}

export interface GithubConfig {
  repositoryName: string;
  orgName?: string;
  since?: string;
  until?: string;
}

export interface GithubRepository {
  commits(githubConfig: GithubConfig): Promise<GithubCommit[]>;
}

export interface GithubService {
  commits(githubConfig: GithubConfig): Promise<GithubEntry[]>;
}

export interface GithubClient {
  commits(githubConfig: GithubConfig): Promise<string[]>;

  getCommitDetails(
    repositoryName: string,
    orgName: string,
    sha: string
  ): Promise<any>;
}
