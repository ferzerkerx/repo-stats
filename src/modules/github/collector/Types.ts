import { CollectorConfig, MetricItem } from '../../../metrics';

export interface GithubMetricItem extends MetricItem {
  sha: string;
  author: string;
  repositoryName: string;
  //TODO
}

export class GithubCollectorConfig implements CollectorConfig {
  repositoryName: string;
  orgName: string;
  since?: string;
  until?: string;

  constructor({ repositoryName, orgName, since = null, until = null }) {
    this.repositoryName = repositoryName;
    this.orgName = orgName;
    this.since = since;
    this.until = until;
  }
}
