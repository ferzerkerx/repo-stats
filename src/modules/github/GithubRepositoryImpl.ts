import {
  GithubClient,
  GithubCommit,
  GithubConfig,
  GithubRepository
} from './Types';
import { Converters } from './Converters';

export class GithubRepositoryImpl implements GithubRepository {
  constructor(private readonly githubClient: GithubClient) {}

  async commits(githubConfig: GithubConfig): Promise<GithubCommit[]> {
    const { repositoryName, orgName } = githubConfig;

    const commitsSha: string[] = await this.githubClient.commits(githubConfig);
    const commitsWithDetailsPromise: Promise<GithubCommit>[] = commitsSha.map(
      commitSha => {
        return this.getCommitDetails(repositoryName, orgName, commitSha);
      }
    );
    return Promise.all(commitsWithDetailsPromise);
  }

  private async getCommitDetails(
    repositoryName: string,
    orgName: string,
    sha: string
  ): Promise<GithubCommit> {
    const commitDetailsResponse = await this.githubClient.getCommitDetails(
      repositoryName,
      orgName,
      sha
    );

    return Converters.toGithubCommit(commitDetailsResponse);
  }
}
