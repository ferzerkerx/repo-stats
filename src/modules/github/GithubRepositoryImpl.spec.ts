import { GithubClient, GithubConfig, GithubRepository } from './Types';
import * as commitsFakeData from './test/commits-response.json';
import * as commitDetailsFakeData from './test/commit-details-response.json';
import { GithubRepositoryImpl } from './GithubRepositoryImpl';

describe('GithubRepository', () => {
  const githubClient: GithubClient = {
    async commits(githubConfig: GithubConfig): Promise<any> {
      return commitsFakeData;
    },
    async getCommitDetails(
      repositoryName: string,
      orgName: string,
      sha: string
    ): Promise<any> {
      return commitDetailsFakeData;
    }
  };

  const githubService: GithubRepository = new GithubRepositoryImpl(
    githubClient
  );

  it('should get commits', async () => {
    const githubConfig: GithubConfig = {
      repositoryName: 'someRepositoryName',
      orgName: 'someOrgName',
      since: '2018-12-03'
    };
    const data = await githubService.commits(githubConfig);
    expect(data).toMatchSnapshot();
  });
});
