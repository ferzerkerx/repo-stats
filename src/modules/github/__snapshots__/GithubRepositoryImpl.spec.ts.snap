// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubRepository should get commits 1`] = `
Array [
  Object {
    "author": "someUser",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "files": Array [
      Object {
        "additions": 0,
        "deletions": 1,
        "filename": "handlers/GENCALC/PartRounder.js",
      },
      Object {
        "additions": 2,
        "deletions": 2,
        "filename": "handlers/file.js",
      },
      Object {
        "additions": 15,
        "deletions": 18,
        "filename": "handlers/file2.js",
      },
    ],
    "linesAdded": 17,
    "linesRemoved": 21,
    "message": "ORG-5917
- some comment",
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "someUser",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "files": Array [
      Object {
        "additions": 0,
        "deletions": 1,
        "filename": "handlers/GENCALC/PartRounder.js",
      },
      Object {
        "additions": 2,
        "deletions": 2,
        "filename": "handlers/file.js",
      },
      Object {
        "additions": 15,
        "deletions": 18,
        "filename": "handlers/file2.js",
      },
    ],
    "linesAdded": 17,
    "linesRemoved": 21,
    "message": "ORG-5917
- some comment",
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "someUser",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "files": Array [
      Object {
        "additions": 0,
        "deletions": 1,
        "filename": "handlers/GENCALC/PartRounder.js",
      },
      Object {
        "additions": 2,
        "deletions": 2,
        "filename": "handlers/file.js",
      },
      Object {
        "additions": 15,
        "deletions": 18,
        "filename": "handlers/file2.js",
      },
    ],
    "linesAdded": 17,
    "linesRemoved": 21,
    "message": "ORG-5917
- some comment",
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "someUser",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "files": Array [
      Object {
        "additions": 0,
        "deletions": 1,
        "filename": "handlers/GENCALC/PartRounder.js",
      },
      Object {
        "additions": 2,
        "deletions": 2,
        "filename": "handlers/file.js",
      },
      Object {
        "additions": 15,
        "deletions": 18,
        "filename": "handlers/file2.js",
      },
    ],
    "linesAdded": 17,
    "linesRemoved": 21,
    "message": "ORG-5917
- some comment",
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "someUser",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "files": Array [
      Object {
        "additions": 0,
        "deletions": 1,
        "filename": "handlers/GENCALC/PartRounder.js",
      },
      Object {
        "additions": 2,
        "deletions": 2,
        "filename": "handlers/file.js",
      },
      Object {
        "additions": 15,
        "deletions": 18,
        "filename": "handlers/file2.js",
      },
    ],
    "linesAdded": 17,
    "linesRemoved": 21,
    "message": "ORG-5917
- some comment",
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
]
`;
